File: frontend/src/games/mental-rotation/core/memory.ts
/* psygames-mental-rotation-memory · VER 1 · 17.09.2026 */
/* psygames-spatial-claude-mac · задача 69f1810f */
/**
 * «ПАМЯТЬ»: ПОКАЗАЛИ — СПРЯТАЛИ — УЗНАЙ ПОВЁРНУТОЙ.
 *
 * Решение Дениса 17.09.2026, дословно: «это твой движок и твоя территория ротации) все
 * упражнения на ментальное вращение используют память) чтобы повернуть в уме надо помнить)».
 * Поэтому режим — не «зрительная память с кубиками», а поворотное задание, в котором эталон
 * приходится держать в голове: варианты появляются ПОСЛЕ того, как фигура исчезла, и они
 * повёрнуты. Удержать образ и повернуть его — одно упражнение, а не два.
 *
 * Задание собирает тот же генератор, что «Поворот» (`buildRotationTask`): те же фигуры, путь
 * поворота, подделки-зеркала и перестановки, проверенные перебором 24 ориентаций. Новое одно —
 * время показа.
 *
 * 🔴 ВРЕМЯ ПОКАЗА — ПАРАМЕТР СТУПЕНИ, А НЕ КОНСТАНТА ЭКРАНА. На нём строится лестница режима
 * (вторая ось — фигура, растёт вместе с уровнем, как у «Поворота»): с 13-го уровня 5 секунд,
 * к 50-му — 1,5. Шаг 100 мс: мельче глаз не различит, а числа остаются читаемыми.
 */
import { buildRotationTask } from './rotation';
import type { MemoryTask, Rng } from './types';

/** Уровень, с которого открывается «Память» (он же в `KIND_UNLOCK`). */
export const MEMORY_FROM_LEVEL = 13;
export const MEMORY_EXPOSURE_START_MS = 5000;
export const MEMORY_EXPOSURE_END_MS = 1500;

/** Время показа эталона на уровне: от 5 с на 13-м до 1,5 с на 50-м, дальше не короче. */
export function memoryExposureMs(level: number): number {
  const шагов = 50 - MEMORY_FROM_LEVEL;
  const k = Math.min(1, Math.max(0, (Math.floor(level) - MEMORY_FROM_LEVEL) / шагов));
  const ms = MEMORY_EXPOSURE_START_MS - (MEMORY_EXPOSURE_START_MS - MEMORY_EXPOSURE_END_MS) * k;
  return Math.round(ms / 100) * 100;
}

export function buildMemoryTask(level: number, rng: Rng): MemoryTask {
  const поворот = buildRotationTask(level, rng);
  return { ...поворот, kind: 'memory', exposureMs: memoryExposureMs(level) };
}

File: flutter/lib/games/mental_rotation/memory.dart
/// «ПАМЯТЬ»: ПОКАЗАЛИ — СПРЯТАЛИ — УЗНАЙ ПОВЁРНУТОЙ — перенос `core/memory.ts`.
///
/// Решение Дениса 17.09.2026, дословно: «все упражнения на ментальное вращение используют
/// память) чтобы повернуть в уме надо помнить)». Поэтому это не «зрительная память с кубиками»,
/// а то же поворотное задание, только эталон исчезает до появления вариантов.
///
/// 🔴 ВРЕМЯ ПОКАЗА — ПАРАМЕТР СТУПЕНИ, А НЕ КОНСТАНТА ЭКРАНА: с 13-го уровня 5 секунд, к 50-му
/// 1,5. Шаг 100 мс — мельче глаз не различит.
library;

import 'rng.dart';
import 'rotation.dart';
import 'task.dart';

/// Уровень, с которого открывается «Память» (он же в карте открытий).
const int memoryFromLevel = 13;
const int memoryExposureStartMs = 5000;
const int memoryExposureEndMs = 1500;

int memoryExposureMs(int level) {
  const steps = 50 - memoryFromLevel;
  final k = ((level - memoryFromLevel) / steps).clamp(0.0, 1.0);
  final ms = memoryExposureStartMs - (memoryExposureStartMs - memoryExposureEndMs) * k;
  return (ms / 100).round() * 100;
}

class MemoryTask implements MentalRotationTask {
  @override
  TaskKind get kind => TaskKind.memory;

  /// «Память» — то же поворотное задание, и верный вариант у неё тот же.
  @override
  int get correctIdx => rotation.correctIdx;

  const MemoryTask({required this.rotation, required this.exposureMs});
  final RotationTask rotation;
  final int exposureMs;
}

MemoryTask buildMemoryTask(int level, Rng rng) =>
    MemoryTask(rotation: buildRotationTask(level, rng), exposureMs: memoryExposureMs(level));
